components/esp_modem/examples/modem_console/main/modem_console_main.cpp

```diff
@@ -289,6 +347,80 @@ extern "C" void app_main(void)
     });
+
+    /* Put the esp32 into deep sleep */
+    const struct DeepSleepArgs {
+        DeepSleepArgs(): timeout(INT1, nullptr, nullptr, "<tout>", "TIMEOUT") {}
+        CommandArgs timeout;
+    } deep_sleep_args;
+    const ConsoleCommand SetDeepSleep("set_deep_sleep", "Put esp32 to deep sleep", &deep_sleep_args, sizeof(deep_sleep_args), [&](ConsoleCommand * c) {
+        int tout = c->get_int_of(&DeepSleepArgs::timeout);
+        ESP_LOGI(TAG, "Entering deep sleep for %d sec", tout);
+        ESP_LOGI(TAG, "Wakeup Cause: %d ", esp_sleep_get_wakeup_cause());
+        esp_deep_sleep(tout * 1000000);
+        return 0;
+    });
+
+    /* Wake up modem */
+    const ConsoleCommand WakeupModem("wakeup_modem", "Wakes up the modem from PSM", no_args, [&](ConsoleCommand * c) {
+        wakeup_modem();
+        ESP_LOGI(TAG, "OK");
+        return 0;
+    });
+
+    /* Enable PSM modem */
+    const ConsoleCommand EnablePSM("enable_psm", "Enables PSM on the modem", no_args, [&](ConsoleCommand * c) {
+        std::string out;
+        CHECK_ERR(dce->at("AT+CPSMS=1", out, 500), ESP_LOGI(TAG, "OK. %s", out.c_str()));
+        return 0;
+    });
+
+    /* Disable PSM modem */
+    const ConsoleCommand DisablePSM("disable_psm", "Disables PSM on the modem", no_args, [&](ConsoleCommand * c) {
+        std::string out;
+        CHECK_ERR(dce->at("AT+CPSMS=0", out, 500),   ESP_LOGI(TAG, "OK. %s", out.c_str()));
+        return 0;
+    });
+
+    /* Get modem PSM cfg */
+    const ConsoleCommand GetModemCfg("get_psm_cfg", "Get PSM config", no_args, [&](ConsoleCommand * c) {
+        std::string out;
+        CHECK_ERR(dce->at("AT+CPSMS?", out, 500),  ESP_LOGI(TAG, "OK. %s", out.c_str()));
+        return 0;
+    });
+
+    /* Set modem PSM config */
+    const struct SetPsmCfgArgs {
+        SetPsmCfgArgs():
+            periodic_tau(STR1, nullptr, nullptr, "<Requested_Periodic-TAU>", "T3412 Timer in 8-bit format"),
+            active_time(STR0, nullptr, nullptr, "<Requested_Active-Time>", "T3324 Timer in 8-bit format"){}
+        CommandArgs periodic_tau;
+        CommandArgs active_time;
+    } set_psm_cfg_args;
+    const ConsoleCommand SetPsmCfg("set_psm_cfg", "Set PSM config", &set_psm_cfg_args, sizeof(set_psm_cfg_args), [&](ConsoleCommand * c) {
+        std::string out;
+        auto periodic_tau = c->get_string_of(&SetPsmCfgArgs::periodic_tau);
+        auto active_time = c->get_string_of(&SetPsmCfgArgs::active_time);
+
+        /* Validate input */
+        if((strlen(periodic_tau.c_str()) != strlen("00000000")) ||
+           (strlen(active_time.c_str()) != strlen("00000000"))) {
+            ESP_LOGE(TAG, "Failed with ERROR: Invalid argument length");
+            return 1;
+        }
+
+        /* Get PSM config */
+        dce->at("AT+CPSMS?", out, 500);
+
+        /* Update PSM config */
+        char set_cmd[35];
+        sprintf(set_cmd, "AT+CPSMS=%c,,,\"%s\",\"%s\"", out.c_str()[8], periodic_tau.c_str(), active_time.c_str());
```

Comment:
Suggest using `asprintf()` or better yet some C++ construct..
```suggestion
        std::string cmd =  "AT+CPSMS=..." +  periodic_tau + "," ...
```

Also, please explain `out.c_str()[8]` -- could we check if `out` really contains 8 bytes?

Reply:
asprintf() needs to have a free in the end, but the next tile CHECK_ERR(... does a return. So there's no scope for a free. That's why I allocated a buffer and used sprint. 
Also, for Sim70xx series the output of AT+CPSMS? is of a fixed length of 35 bytes.

Reply:
Done

Reply:
Ok, Thanks for updates,

Could you also please check if `out.size() > 8`, just to be sure we could dereference `out.c_str()[8]`

Reply:
Done